sw/device/rom_exts/manifest.md

```diff
@@ -184,13 +184,35 @@ Notes:
     This is used when signing and validating the image. This happens in the
     Mask ROM, as well as during firmware update.
 
-1.  **Usage Constraints** This is a 32-bit enumeration value which denotes what
-    kinds of devices the ROM_EXT image may be used on.
-
-    This is used by the Mask ROM to validate the image is running on the right
-    kind of hardware.
-
-    **Open Q** What possible values should this have?
+1.  **Usage Constraints** This denotes which pieces of the OpenTitan Device
+    Identifier should also be added to the digest of this image. The OpenTitan
+    Device Identifier is a 256-bit value provisioned into the device's OTP
+    memory early in manufacturing.
+
+    This allows the ROM_EXT author to constrain a ROM_EXT to a single device or
+    a group of devices, such that the ROM_EXT will not validate (and therefore
+    not boot) on devices not in that group. This can use the Silicon Creator
+    identifier, the product identifier, the individual device identifier, or any
+    data from the SKU-specific device identifier.
+
+    This is a 32-bit unsigned numeric value.
+
+    Each bit in the usage constraint value denotes a byte in the 256bit Device
+    Identifier. If that bit is set, then that byte in the should be added to the
+    end of the digest, otherwise a zero-byte should be added to the end of the
+    digest. The `i`th bit (counting from the least significant bit) corresponds
+    to the `i`th byte of the OpenTitan Device Identifier.
+
+    ```
+    add_to_digest(rom_ext_image)
+    for (i in 0..31) {
+      if (usage_constraint_value & (1 << i))
+        add_to_digest((uint8_t)otp_get_device_id_byte(i))
+      else
+        add_to_digest((uint8_t)0b00000000)
```

Comment:
nit: Prefer a value with high hamming distance

Reply:
Ok, so something like `0xAA`?

Reply:
Whatever we choose here should probably have to be reserved out of the allocated creator, product, and device-specific identifiers, or else you could end up validating something on a specific device that you did not intend to validate. 

i.e. we need to ensure that `masked(device id A, less specific mask)` is never equal to `masked(device id B, more specific mask)` for any two device ids, which would entail reserving all IDs that contain a byte that matches the value chosen here.

Reply:
I've been thinking about this more, and I think the way to solve this is the following:
- Reserve all values where any single byte is 0x00
- Implement the following for checking the usage constraint, written in some kind of pseudo-code:

```
device_id_masked = [manifest_usage_constraint.get_bit(i) ? device_id.get_byte(i) : 0x00 for i in 0..(256/8)]
usage_constraint_value = [device_id_masked.get_byte(i) ^ 0xAA for i in 0..(256/8)]

actual_hash = HASH([rom_ext_image || usage_constaint_value])
expected_hash = get_digest_from_signature(pub_key, signature)

if (actual_hash == expected_hash)
  // boot successfully
else
  // boot error
```

This I think solves the issue i pointed to above about more and less specific ids, and also solves the high hamming distance request from @moidx.

Reply:
What you had originally was fine. I was just asking to change the else to hash with a non 0s or 1s value to defend against FI attacks. Also I still find the pseudocode confusing by just referencing the device_id. I am expecting it to me more flexible and over potentially cover other fuse offsets once we define the OTP memory map.

Also, It is most likely okay to define fuses with uint32_t alignment. So your calculation can be done in uint32_t offsets.

```
    add_to_digest(rom_ext_image)
    for (i in 0..31) {
      if (usage_constraint_value & (1 << i))
        add_to_digest((uint32_t)otp_get_word(i))
      else
        add_to_digest((uint32_t)0x5A5A5A5A)
```

Reply:
We will revisit this conversation later during the system state module definition, but the issues are noted.